### ISSUE TITLE
Are the AUC/ROC etc. images based on the train set, validation set(s) or a mix of all?

### ISSUE THREAD

**Issue by tfgstudios** [mljar/mljar-supervised#533]
Hey, thanks for this lib. I was surprised something like this exists. It helps me a lot in selecting features, maybe I'll even use it in production.

Basically, I'd like to know if the analysis in the rendered images estimate the model's 'true' performance (i.e. on data it has never seen before) or just the train performance (i.e. ability to _at least_ overfit).
The way I understand it, you always use a validation mechanism, be it via cv or train-test split. Can you confirm if the above-mentioned metrics will at least approach the performance on a _test_ set, i.e. completely external data?

My data has 9k columns (flatten timeseries of 150 timesteps * 13 features), multi-classification problem.
I get a ~0.8 AUC on the Precision-Recall curve and a ROC of ~0.85 which is suspiciously high for my problem complexity.

Thanks

**Comment by pplonski** [comment]
Hi @tfgstudios,

I'm glad that you found the package and it works for you.

All metrics are computed on validation part of the data. Does it guarantee that models will work in the same way on test data? It depends on how the new data will be similar to training data. If both (train and test) will have the similar distribution, then it will work (most cases).

Time-series data can be tricky. You can easily do information leakage in data transformation. So you will end up with overfitted model.

**Comment by tfgstudios** [comment]
> Hi @tfgstudios,
> 
> I'm glad that you found the package and it works for you.
> 
> All metrics are computed on validation part of the data. Does it guarantee that models will work in the same way on test data? It depends on how the new data will be similar to training data. If both (train and test) will have the similar distribution, then it will work (most cases).
> 
> Time-series data can be tricky. You can easily do information leakage in data transformation. So you will end up with overfitted model.

Thank you for replying.
Yes, I've spend quite some time finding ways in which to transform the data without info leakage between the sets. 
I'm thankful for any pointers you have for my problem space. What I'm doing is: 
Each x is a list of _flattened_ features: a0, b0, c0.... a9, b9, c9;
The digit represents the timestep (10 timesteps for simplicity). a,b,c are the _actual_ features.
y is a label in {0, 1, 2}
I didn't transform the data at all at the set level (just clipped some outliers), but instead on the _sequence_ level.
What I did, for each _actual_ feature (a, b, c), I took the return relative to the 1st timestep (because I don't feel that using the common approach of pandas' `[pct_change()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pct_change.html)` captures enough temporal information), i.e.
a0=0
a1=a1_real_value / a0_real_value
a2=a2_real_value / a0_real_value
etc. for each feature, which gives me 2 things: more stationarity and easier comparison of different time-steps within a sequence (which the `pct_change()` method doesn't provide as getting the value of any time step requires you to go through all the steps before it which IMO is a very indirect way of specifying values) I didn't test if my approach is better than the commonly used one, but I'll eventually do both.

Then just because you require a (n_steps, n_features) input shape, I flatten the timesteps into what your algo sees as separate different features.

Now, after 15 h of training on Compete, it found some good results, but on testing the confusion matrix is just random chance. The train-test set was split 80/20, nothing fancy. 
Here are the params:
`{
    "mode": "Compete",
    "ml_task": "multiclass_classification",
    "results_path": "AutoML_6",
    "total_time_limit": 54000,
    "model_time_limit": null,
    "algorithms": [
        "Random Forest",
        "Xgboost",
        "Decision Tree",
        "Baseline",
        "LightGBM",
        "Extra Trees",
        "CatBoost",
        "Neural Network",
        "Nearest Neighbors"
    ],
    "train_ensemble": true,
    "stack_models": false,
    "eval_metric": "logloss",
    "validation_strategy": {
        "validation_type": "split",
        "train_ratio": 0.9,
        "shuffle": true,
        "stratify": true,
        "X_path": "AutoML_6\\X.data",
        "y_path": "AutoML_6\\y.data",
        "results_path": "AutoML_6",
        "sample_weight_path": "AutoML_6\\sample_weight.data"
    },
    "verbose": 1,
    "explain_level": 2,
    "golden_features": true,
    "features_selection": true,
    "start_random_models": 10,
    "hill_climbing_steps": 2,
    "top_models_to_improve": 3,
    "boost_on_errors": false,
    "kmeans_features": true,
    "mix_encoding": true,
    "max_single_prediction_time": null,
    "n_jobs": -1,
    "random_state": 1234,
    "saved": [
        "1_Baseline",
        "2_DecisionTree",
        "3_DecisionTree",
        "4_DecisionTree",
        "5_Default_LightGBM",
        "6_Default_Xgboost",
        "7_LightGBM",
        "12_DecisionTree",
        "13_DecisionTree",
        "14_DecisionTree",
        "Ensemble"
    ],
    "fit_level": "finished",
    "best_model": "Ensemble",
    "load_on_predict": [
        "5_Default_LightGBM",
        "6_Default_Xgboost",
        "7_LightGBM",
        "Ensemble"
    ]
}`
Confusion matrix (which looks very close to what I want, class 0 being not so important):
![confusion_matrix_normalized](https://user-images.githubusercontent.com/5688727/162453785-41d01823-81d1-4d93-9b32-c02a84a1d27a.png)

But on the test set it basically predicts no better than randomly (respecting the distribution of classes, as this is a slightly imbalanced set where class 0 appears more often than 1,2)

Any kind of pointer would be highly appreciated.

**Comment by pplonski** [comment]
@tfgstudios instead of shifted values in the additional columns, you can try to use % change between samples and can try some moving window features.
